Possibility of not specifying the persistency manager

paolafer opened the pull request:
This PR removes the need of specifying a persistency manager to run a simulation.
It can be useful for debugging, an example of which has been added in a pair of macros.
Additionally, the names of two configuration parameters of the `PersistencyManager` class have been changed to follow the current convention.

paolafer committed:
Eliminate the need of specifying a persistency manager

paolafer committed:
Add example of not using persistency manager

paolafer committed:
Remove unused variable

paolafer committed:
Add default name for output file to avoid crash if not specified

paolafer committed:
Use new variable name in the persistency manager

paolafer committed:
Modify automatic tests with new variable name

paolafer committed:
Update example macros with new variable names

gonzaponte commented on macros/NEXT100_debug.config.mac:
```diff
@@ -0,0 +1,25 @@
+/Generator/SingleParticle/min_energy 1 eV
+/Generator/SingleParticle/max_energy 1 eV
+/Generator/SingleParticle/region AD_HOC
+/Generator/SingleParticle/momentum 0 0 1
```
add a newline at the end of file

gonzaponte replied:
Did you try to shoot something that produced hits? Before I saw this PR existed I started developing something similar in parallel and had a segfault because it attempted to store trajectories.

Is it meant to be used only for non-interacting events or is it more general?
If it is the latter, I suggest shooting an electron in the example macro to cover all cases. It only takes a few milliseconds to run anyway.

gonzaponte commented on macros/nexus_example1.config.mac:
```diff
@@ -19,4 +19,4 @@
 
-/nexus/persistency/outputFile nexus.next
\ No newline at end of file
+/nexus/persistency/output_file nexus.next
```
add newline at the end of file

gonzaponte commented on source/base/NexusApp.h:
```diff
@@ -55,6 +55,8 @@ namespace nexus {
     G4String trkact_name_; ///< Name of the chosen tracking action
     G4String stkact_name_; ///< Name of the chosen stacking action
 
+    G4bool pman_; ///< True if the persistency manager is set
+
```
I don't know what's the convention in nexus, but usually in C++ you could simply check if `pm_` is a `nullptr`.

I like that there is a more explicit variable, though, so for me this is fine.

paolafer replied:
If you think this way it is easier to understand what the code does, then I would keep it.

gonzaponte commented on source/persistency/PersistencyManager.cc:
```diff
@@ -42,15 +42,15 @@ REGISTER_CLASS(PersistencyManager, PersistencyManagerBase)
 
 PersistencyManager::PersistencyManager():
-  PersistencyManagerBase(), msg_(0), ready_(false),
+PersistencyManagerBase(), msg_(0), output_file_("nexus_out"), ready_(false),
```
why does it crash if you don't specify a default value?

paolafer replied:
In fact, it doesn't: it doesn't produce an output file at all.

gonzaponte reviewed (commented):
Looks good to me. Just a couple of questions below.

soleti commented on source/base/NexusApp.cc:
```diff
@@ -105,13 +105,13 @@ NexusApp::NexusApp(G4String init_macro): G4RunManager(), gen_name_(""),
 
- // PersistencyManager::Initialize(init_macro, macros_, delayed_);
+  // Set the persistency manager, if needed
+  if (pm_name_ != "") {
```
I think `!pm_name.empty()` is slightly faster than `!= ""` and the "recommended" way

soleti reviewed (commented):
Looks good to me (although I have limited experience with nexus)

paolafer committed:
Add new lines at the end of the file and turn off verbosity

The particle has been changed from geantino to electron to cover a
more general case of interaction with the detector materials.

paolafer committed:
Use the recommended way of checking if a string is empty
